# Проблема собственных значений матрицы

Выполнил студент 3 курса СЦТ Лютарь Владислав

In [26]:
import numpy as np

def Norm2(v):
    return sum([v_[0]*v_[0] for v_ in v])**.5

### генератор матрицы

генерить буду матрицы вида:
$$ \begin{pmatrix}
1 & 1 & 1 & ... & 1 \\
1 & 2 & 2 & ... & 2 \\
& & ... \\
1 & 2 & 3 & ... & n
\end{pmatrix} $$

У матриц такого вида есть прикол, что их определители равны 1. Это крайне просто проверяется если использовать базовые свойства определителя.

In [27]:
n = 4

A = np.zeros((n,n))

for i in range(n):
    for j in range(n):
        if i<=j: 
            A[i, j] = i+1
            A[j, i] = i+1

A, np.linalg.eig(A).eigenvalues

(array([[1., 1., 1., 1.],
        [1., 2., 2., 2.],
        [1., 2., 3., 3.],
        [1., 2., 3., 4.]]),
 array([8.29085937, 1.        , 0.42602205, 0.28311858]))

## Метод итераций

In [28]:
def IterMethod(A, x0_, eps):
    kaunt = 0

    x0 = x0_ / x0_.reshape((1,n)).max()

    xnext = A @ x0
    xprev = np.zeros((n,1))

    while Norm2(xnext - xprev) > eps:
        xprev = xnext
        xnext = A @ (xnext / xnext.reshape((1,n)).max())
        kaunt += 1

    return xnext.reshape((1,n)).max(), kaunt

# A = np.array([
#     [2, 0, 0],
#     [0, 3, 0],
#     [0, 0, 4]
# ])

# X0 = np.array([1 for i in range(n)]).reshape((n,1))
# IterMethod(A, X0, 0.001)

In [29]:
print("ε = 0.001")
for n in range(2,10):

    A = np.zeros((n,n))

    for i in range(n):
        for j in range(n):
            if i<=j: 
                A[i, j] = i+1
                A[j, i] = i+1

    X0 = np.array([1 for _ in range(n)]).reshape((n,1))
    print(max(np.linalg.eig(A).eigenvalues)," | ",IterMethod(A, X0, 0.001))


print("\nε = 0.0000001")
for n in range(2,10):

    A = np.zeros((n,n))

    for i in range(n):
        for j in range(n):
            if i<=j: 
                A[i, j] = i+1
                A[j, i] = i+1

    X0 = np.array([1 for _ in range(n)]).reshape((n,1))
    print(max(np.linalg.eig(A).eigenvalues)," | ",IterMethod(A, X0, 0.0000001))

ε = 0.001
2.618033988749895  |  (2.6180555555555554, 5)
5.048917339522298  |  (5.048951048951049, 5)
8.290859369381591  |  (8.290909090909091, 5)
12.34353751967705  |  (12.343607372773377, 5)
17.20685726740093  |  (17.206868168248207, 6)
22.880782741943573  |  (22.880796783857708, 6)
29.36529789437195  |  (29.365315534632153, 6)
36.660394667024335  |  (36.660416360517885, 6)

ε = 0.0000001
2.618033988749895  |  (2.6180339985218035, 9)
5.048917339522298  |  (5.0489173484443475, 9)
8.290859369381591  |  (8.290859370657467, 10)
12.34353751967705  |  (12.343537521240004, 10)
17.20685726740093  |  (17.206857269347278, 10)
22.880782741943573  |  (22.880782744354367, 10)
29.36529789437195  |  (29.365297897322392, 10)
36.660394667024335  |  (36.660394670586946, 10)


In [30]:
# бонус

n=1000

A = np.zeros((n,n))

for i in range(n):
    for j in range(n):
        if i<=j: 
            A[i, j] = i+1
            A[j, i] = i+1

X0 = np.array([1 for _ in range(n)]).reshape((n,1))
print(max(np.linalg.eig(A).eigenvalues)," | ",IterMethod(A, X0, 0.0000001))

405690.2039584478  |  (405690.2039584477, 16)


## Метод Якоби

In [31]:
# n = 3

# A = np.zeros((n,n))

# for i in range(n):
#     for j in range(n):
#         if i<=j: 
#             A[i, j] = i+1
#             A[j, i] = i+1
# print(np.linalg.eig(A).eigenvalues)


# def IndexOfMaxUpper(A):
#     mx = 0
#     idx = []
#     n = len(A)

#     for i in range(n):
#         for j in range(n):
#             if i<j: 
#                 if mx < abs(A[i, j]):
#                     idx = (i, j)
#                     mx = A[i, j]
#     return idx


# def StopCondition(A, eps):
#     Sum = 0
#     for i in range(n):
#         for j in range(n):
#             if i != j:
#                 Sum += A[i, j]*A[i, j]
#     return Sum > eps

# def CondSum(A):
#     Sum = 0
#     n = len(A)
#     for i in range(n):
#         for j in range(n):
#             if i != j:
#                 Sum += A[i, j]*A[i, j]
#     return Sum


# def JacobiMethod(A_, eps):
#     A = A_
#     n = len(A)
#     bl = np.zeros((1, n))[0]
#     bm = np.zeros((1, n))[0]

#     kaunt = 0

#     while CondSum(A) > eps:
#         kaunt += 1

#         kortezh = IndexOfMaxUpper(A)
#         #print(kortezh)
#         l = kortezh[0]; m = kortezh[1]

#         if A[l, l] == A[m, m]:
#             phi = np.pi/4
#         else:
#             phi = np.arctan(2*A[l, m] / (A[l, l] - A[m, m])) / 2
            
#         s = np.sin(phi)
#         c = np.cos(phi)

#         A_ll = A[l, l]
#         A_lm = A[l, m]
#         A_mm = A[m, m]

#         A[l, l] = c*c * A_ll + 2*c*s * A_lm + s*s * A_mm
#         A[m, m] = s*s * A_ll - 2*c*s * A_lm + c*c * A_mm
#         A[l, m] = 0
#         A[m, l] = 0

#         for i in range(n):
#             if (i != l) and (i != m):
#                 A_li = A[l, i]
#                 A_mi = A[m, i]

#                 A[l, i] = c * A_li - s * A_mi
#                 A[m, i] = s * A_li + c * A_mi

#                 A[i, l] = A[l, i]
#                 A[i, m] = A[m, i]
#         print(A, "\n")

#     return A, kaunt

#JacobiMethod(A, 0.001)


In [35]:
# n = 3

# A = np.zeros((n,n))

# for i in range(n):
#     for j in range(n):
#         if i<=j: 
#             A[i, j] = i+1
#             A[j, i] = i+1
# print(np.linalg.eig(A).eigenvalues)


def IndexOfMaxUpper(A):
    mx = 0
    idx = []
    n = len(A)

    for i in range(n):
        for j in range(n):
            if i<j: 
                if mx < abs(A[i, j]):
                    idx = (i, j)
                    mx = A[i, j]
    return idx


# def StopCondition(A, eps):
#     Sum = 0
#     for i in range(n):
#         for j in range(n):
#             if i != j:
#                 Sum += A[i, j]*A[i, j]
#     return Sum > eps

# def CondSum(A):
#     Sum = 0
#     n = len(A)
#     for i in range(n):
#         for j in range(n):
#             if i != j:
#                 Sum += A[i, j]*A[i, j]
#     return Sum


def JacobiMethod(A_, eps):
    A = A_
    n = len(A)
    bl = np.zeros((1, n))[0]
    bm = np.zeros((1, n))[0]

    kaunt = 0

    #while np.sum(A ** 2) - np.sum(np.diag(A ** 2)) > eps:
    for __ in range(2000000):
        kaunt += 1

        kortezh = IndexOfMaxUpper(A)
        #print(kortezh)
        l = kortezh[0]; m = kortezh[1]

        if A[l, l] == A[m, m]:
            phi = np.pi/4
        else:
            phi = np.arctan2(2*A[l, m],(A[l, l] - A[m, m])) / 2
            
        s = np.sin(phi)
        c = np.cos(phi)

        A_ll = A[l, l]
        A_lm = A[l, m]
        A_mm = A[m, m]

        A[l, m] = 0
        A[m, l] = 0

        for i in range(n):
            A_li = A[l, i]
            A_mi = A[m, i]

            if (i != l) and (i != m):
                b_li = c * A_li + s * A_mi
                b_mi = -s * A_li + c * A_mi

                A[l, i] = b_li
                A[m, i] = b_mi
                A[i, l] = b_li
                A[i, m] = b_mi

            elif i==l:
                b_ll = c*c * A_ll + 2*c*s * A_lm + s*s * A_mm
                A[l, l] = b_ll

            elif i==m:
                b_lm = c*s * (A_mm - A_ll) + (c*c - s*s) * A_lm
                b_mm = s*s * A_ll - 2*c*s * A_lm + c*c * A_mm

                A[l, m] = b_lm
                A[m, l] = b_lm
                A[m, m] = b_mm


        # A_ll = A[l, l]
        # A_lm = A[l, m]
        # A_mm = A[m, m]

        # A[l, l] = c*c * A_ll + 2*c*s * A_lm + s*s * A_mm
        # A[m, m] = s*s * A_ll - 2*c*s * A_lm + c*c * A_mm
        # A[l, m] = 0
        # A[m, l] = 0

        # for i in range(n):
        #     if (i != l) and (i != m):
        #         A_li = A[l, i]
        #         A_mi = A[m, i]

        #         A[l, i] = c * A_li - s * A_mi
        #         A[m, i] = s * A_li + c * A_mi

        #         A[i, l] = A[l, i]
        #         A[i, m] = A[m, i]

        if np.sum(A ** 2) - np.sum(np.diag(A ** 2)) < eps: break
                
        #print([A[_,_] for _ in range(n)], "\n")

    return [A[_,_] for _ in range(n)], kaunt

#JacobiMethod(A, 0.001)


In [36]:
n = 3

A = np.zeros((n,n))

for i in range(n):
    for j in range(n):
        if i<=j: 
            A[i, j] = i+1
            A[j, i] = i+1


print(np.linalg.eig(A).eigenvalues.round(7))
JacobiMethod(A, 0.0001)

[5.0489173 0.6431041 0.3079785]


([5.048221468528735, 0.6433861490772014, 0.30839238239406497], 2000000)

In [34]:
A = np.array([
    [3,1,1],
    [1,3,2],
    [1,2,3]
])


print(np.linalg.eig(A).eigenvalues.round(7))
JacobiMethod(A, 0.0001)

[2.2679492 5.7320508 1.       ]


([5, 2, 1], 2)

## Ну наверное резюмирование

Метод итераций работает безупречно и сходится вполне себе быстро, кодится легко и вщ лапочка

Метод вращений по какой-то причине либо тупо не может сойтись, либо сходится к чему-то неправильному. Выше я несколько раз пробовал переписывать алгоритм и результат лучше не стал. 

Еще я пробовал выводить выводить полученные сз на каждой итерации и заметил, что они давным давно сошлись к [5.048221468528735, 0.6433861490772014, 0.30839238239406497]. Затем я попробовал вывести значение суммы из условия и оказалось что сумма вообще сходится к какому-то ненулевому конечному числу. По этой причине сойтись он и не может. Такие вот дела :|